### PR TITLE
ports /tg/ status effects

### PR DESCRIPTION
**Pull request by spookydonut**
## About The Pull Request
This ports tg status effects

## Why It's Good For The Game
Instead of running in life these process independently. 
Also allows arbitrary status effects to be added. 

Currently progress:
- [x] stuns
- [x] knockdown
- [x] knockedout
- [ ] frozen
- [x] sleeping
- [x] unconscious
- [ ] paralyzed
- [x] confused

frozen and paralyzed will need mobility flags

### CODE REVIEW

File: code/__DEFINES/components.dm [hunk 1]
```diff
@@ -217,6 +217,17 @@
...
 #define COMSIG_LIVING_SET_CANMOVE "living_set_canmove"			//from base of /mob/living/set_canmove(): (canmove)
 
+//ALL OF THESE DO NOT TAKE INTO ACCOUNT WHETHER AMOUNT IS 0 OR LOWER AND ARE SENT REGARDLESS!
+#define COMSIG_LIVING_STATUS_STUN "living_stun"					//from base of mob/living/Stun() (amount, update, ignore)
+#define COMSIG_LIVING_STATUS_KNOCKDOWN "living_knockdown"		//from base of mob/living/Knockdown() (amount, update, ignore)
+#define COMSIG_LIVING_STATUS_PARALYZE "living_paralyze"			//from base of mob/living/Paralyze() (amount, update, ignore)
+#define COMSIG_LIVING_STATUS_IMMOBILIZE "living_immobilize"		//from base of mob/living/Immobilize() (amount, update, ignore)
+#define COMSIG_LIVING_STATUS_UNCONSCIOUS "living_unconscious"	//from base of mob/living/Unconscious() (amount, update, ignore)
+#define COMSIG_LIVING_STATUS_SLEEP "living_sleeping"			//from base of mob/living/Sleeping() (amount, update, ignore)
+	#define COMPONENT_NO_STUN 1			//For all of them
```
Comment (Rohesie) [inline code/__DEFINES/components.dm]:
```suggestion
	#define COMPONENT_NO_STUN (1<<0)			//For all of them
```
Flag patterns.

File: code/__DEFINES/components.dm [hunk 2]
```diff
@@ -225,6 +225,18 @@
...
+#define COMSIG_LIVING_STATUS_STUN "living_stun"					//from base of mob/living/Stun() (amount, update, ignore)
+#define COMSIG_LIVING_STATUS_KNOCKDOWN "living_knockdown"		//from base of mob/living/Knockdown() (amount, update, ignore)
+#define COMSIG_LIVING_STATUS_PARALYZE "living_paralyze"			//from base of mob/living/Paralyze() (amount, update, ignore)
+#define COMSIG_LIVING_STATUS_IMMOBILIZE "living_immobilize"		//from base of mob/living/Immobilize() (amount, update, ignore)
+#define COMSIG_LIVING_STATUS_UNCONSCIOUS "living_unconscious"	//from base of mob/living/Unconscious() (amount, update, ignore)
+#define COMSIG_LIVING_STATUS_SLEEP "living_sleeping"			//from base of mob/living/Sleeping() (amount, update, ignore)
+#define COMSIG_LIVING_STATUS_CONFUSED "living_confused"			//from base of mob/living/Confused() (amount, update, ignore)
+	#define COMPONENT_NO_STUN 1			//For all of them
+#define COMSIG_LIVING_CAN_TRACK "mob_cantrack"					//from base of /mob/living/can_track(): (mob/user)
+	#define COMPONENT_CANT_TRACK 1
```
Comment (Rohesie) [inline code/__DEFINES/components.dm]:
```suggestion
	#define COMPONENT_CANT_TRACK (1<<0)
```

File: code/datums/status_effects/status_effect.dm [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+	on_creation(arglist(arguments))
+
+/datum/status_effect/proc/on_creation(mob/living/new_owner, ...)
+	if(new_owner)
+		owner = new_owner
+	if(owner)
+		LAZYADD(owner.status_effects, src)
+	if(!owner || !on_apply())
+		qdel(src)
+		return
```
Comment (Rohesie) [inline code/datums/status_effects/status_effect.dm]:
```suggestion
		CRASH("Status effect [type] created improperly. Owner: [isnull(owner) "null" : "[owner]"].")
```

File: code/modules/mob/living/living_defines.dm [hunk 4]
```diff
@@ -69,6 +64,10 @@
 
 	var/tinttotal = TINT_NONE
 
+	var/list/status_effects //a list of all status effects the mob has
+
+	var/stun_absorption = null //converted to a list of stun absorption sources this mob has when one is added
```
Comment (Rohesie) [inline code/modules/mob/living/living_defines.dm]:
```suggestion
	var/list/stun_absorption = null //converted to a list of stun absorption sources this mob has when one is added
```

File: code/modules/mob/living/living_health_procs.dm [hunk 5]
```diff
@@ -75,14 +75,11 @@
...
-				visible_message("<span class='warning'>\The [src] slumps to the ground, too weak to continue fighting.</span>",
-					"<span class='warning'>You slump to the ground, you're too exhausted to keep going...</span>")
-			knock_down(4)
-		if(1 to 3)
-			set_knocked_down(4, FALSE)
+	if(!IsKnockdown())
+		if(feedback)
+			visible_message("<span class='warning'>\The [src] slumps to the ground, too weak to continue fighting.</span>",
+				"<span class='warning'>You slump to the ground, you're too exhausted to keep going...</span>")
+	Knockdown(80)
```
Comment (Rohesie) [inline code/modules/mob/living/living_health_procs.dm]: Perhaps use `AmountKnockdown` here? The idea is to keep the mob down until the stamina regens, but not long past that.

File: tgui-next/node_modules/tgui [hunk 6]
```diff
@@ -0,0 +1 @@
+C:/Users/ash/Documents/git/TerraGov-Marine-Corps/tgui-next/packages/tgui
```
Comment (Rohesie) [inline tgui-next/node_modules/tgui]: Unintended change?

File: code/__DEFINES/components.dm (hunk 2 above)
Comment (spookydonut) [reply code/__DEFINES/components.dm]: these arent being used prettu sure

File: code/datums/status_effects/status_effect.dm (hunk 3 above)
Comment (spookydonut) [reply code/datums/status_effects/status_effect.dm]: This is from tg, if they dont have this issue its highly likely we wont either

File: code/modules/mob/living/living_health_procs.dm (hunk 5 above)
Comment (spookydonut) [reply code/modules/mob/living/living_health_procs.dm]: Knockdown(x) is knockdown = min(x, currentknockdown)

Comment (spookydonut) [reply code/modules/mob/living/living_health_procs.dm]: This works exactly as it did before/

File: code/datums/status_effects/status_effect.dm (hunk 3 above)
Comment (Rohesie) [reply code/datums/status_effects/status_effect.dm]: The suggestion is just making a silent error not be silent. It's free, and if it's never triggered, that is fine, but then the check shouldn't be there in the first place.

File: code/modules/mob/living/living_health_procs.dm (hunk 5 above)
Comment (Rohesie) [reply code/modules/mob/living/living_health_procs.dm]: Ah, didn't notice `Knockdown(80)` is not inside the `if(!IsKnockdown())` check.
Alright.

File: code/datums/status_effects/status_effect.dm (hunk 3 above)
Comment (spookydonut) [reply code/datums/status_effects/status_effect.dm]: I'm reasonable certain theres cases where it happens as a normal part of status effects and shouldnt error

File: code/datums/status_effects/debuffs.dm [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+	tick_interval = 0
+	status_type = STATUS_EFFECT_REPLACE
+	alert_type = null
+	var/needs_update_stat = FALSE
+
+/datum/status_effect/incapacitating/on_creation(mob/living/new_owner, set_duration, updating_canmove)
+	if(isnum(set_duration))
+		duration = set_duration
+	. = ..()
+	if(.)
```
Comment (Rohesie) [inline code/datums/status_effects/debuffs.dm]: Early return.

File: code/datums/status_effects/debuffs.dm [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+	name = "Asleep"
+	desc = "You've fallen asleep. Wait a bit and you should wake up. Unless you don't, considering how helpless you are."
+	icon_state = "asleep"
+
+//ADMIN SLEEP
+/datum/status_effect/incapacitating/adminsleep
+	id = "adminsleep"
+	alert_type = /obj/screen/alert/status_effect/adminsleep
+	needs_update_stat = TRUE
+	duration = -1
```
Comment (Rohesie) [inline code/datums/status_effects/debuffs.dm]: Define.

File: code/datums/status_effects/debuffs.dm [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+	if(owner.maxHealth)
+		var/health_ratio = owner.health / owner.maxHealth
+		var/healing = -0.2
+		if((locate(/obj/structure/bed) in owner.loc))
+			healing -= 0.3
+		else if((locate(/obj/structure/table) in owner.loc))
+			healing -= 0.1
+		for(var/obj/item/bedsheet/bedsheet in range(owner.loc,0))
+			if(bedsheet.loc != owner.loc) //bedsheets in your backpack/neck don't give you comfort
+				continue
```
Comment (Rohesie) [inline code/datums/status_effects/debuffs.dm]: This is a meme. Please remove it.

File: code/datums/status_effects/debuffs.dm [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+/datum/status_effect/incapacitating/unconscious/tick()
+	if(owner.getStaminaLoss())
+		owner.adjustStaminaLoss(-0.3) //reduce stamina loss by 0.3 per tick, 6 per 2 seconds
+
+//SLEEPING
+/datum/status_effect/incapacitating/sleeping
+	id = "sleeping"
+	alert_type = /obj/screen/alert/status_effect/asleep
+	needs_update_stat = TRUE
+	var/mob/living/carbon/carbon_owner
```
Comment (Rohesie) [inline code/datums/status_effects/debuffs.dm]: Use a flag instead. Something like `OWNER_CAN_DREAM`

File: code/datums/status_effects/status_effect.dm [hunk 11]
```diff
@@ -0,0 +1,243 @@
+//Status effects are used to apply temporary or permanent effects to mobs. Mobs are aware of their status effects at all times.
+//This file contains their code, plus code for applying and removing them.
+//When making a new status effect, add a define to status_effects.dm in __DEFINES for ease of use!
+
+/datum/status_effect
+	var/id = "effect" //Used for screen alerts.
+	var/duration = -1 //How long the status effect lasts in DECISECONDS. Enter -1 for an effect that never ends unless removed through some means.
```
Comment (Rohesie) [inline code/datums/status_effects/status_effect.dm]: Define.

File: code/datums/status_effects/status_effect.dm [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+	if(original_duration == -1)
+		return
+	duration = world.time + original_duration
+
+//clickdelay/nextmove modifiers!
+/datum/status_effect/proc/nextmove_modifier()
+	return 1
+
+/datum/status_effect/proc/nextmove_adjust()
+	return 0
```
Comment (Rohesie) [inline code/datums/status_effects/status_effect.dm]: These are unused both here and in /tg/
Not really interesting to have.

File: code/datums/status_effects/status_effect.dm [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+	LAZYINITLIST(status_effects)
+	for(var/datum/status_effect/S in status_effects)
+		if(S.id == initial(S1.id) && S.status_type)
+			if(S.status_type == STATUS_EFFECT_REPLACE)
+				S.be_replaced()
+			else if(S.status_type == STATUS_EFFECT_REFRESH)
+				S.refresh()
+				return
+			else
+				return
```
Comment (Rohesie) [inline code/datums/status_effects/status_effect.dm]: Early continues.

File: code/datums/status_effects/status_effect.dm [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
+	var/list/arguments = args.Copy()
+	arguments[1] = src
+	S1 = new effect(arguments)
+	. = S1
```
Comment (Rohesie) [inline code/datums/status_effects/status_effect.dm]:
```suggestion
	return S1
```

File: code/datums/status_effects/status_effect.dm [hunk 15]
```diff
@@ -0,0 +1,243 @@
...
+
+/mob/living/proc/remove_status_effect(effect) //removes all of a given status effect from this mob, returning TRUE if at least one was removed
+	. = FALSE
+	if(status_effects)
+		var/datum/status_effect/S1 = effect
+		for(var/datum/status_effect/S in status_effects)
+			if(initial(S1.id) == S.id)
+				qdel(S)
+				. = TRUE
```
Comment (Rohesie) [inline code/datums/status_effects/status_effect.dm]: Early returns and less nesting.

File: code/datums/status_effects/status_effect.dm [hunk 16]
```diff
@@ -0,0 +1,243 @@
...
+
+/mob/living/proc/has_status_effect(effect) //returns the effect if the mob calling the proc owns the given status effect
+	. = FALSE
+	if(status_effects)
+		var/datum/status_effect/S1 = effect
+		for(var/datum/status_effect/S in status_effects)
+			if(initial(S1.id) == S.id)
+				return S
```
Comment (Rohesie) [inline code/datums/status_effects/status_effect.dm]: Early returns and less nesting.

File: code/datums/status_effects/status_effect.dm [hunk 17]
```diff
@@ -0,0 +1,243 @@
...
+
+/mob/living/proc/has_status_effect_list(effect) //returns a list of effects with matching IDs that the mod owns; use for effects there can be multiple of
+	. = list()
+	if(status_effects)
+		var/datum/status_effect/S1 = effect
+		for(var/datum/status_effect/S in status_effects)
+			if(initial(S1.id) == S.id)
+				. += S
```
Comment (Rohesie) [inline code/datums/status_effects/status_effect.dm]: Early returns and less nesting.

File: code/datums/status_effects/status_effect.dm [hunk 18]
```diff
@@ -0,0 +1,243 @@
...
+/datum/status_effect/stacking/proc/can_have_status()
+	return owner.stat != DEAD
+
+/datum/status_effect/stacking/proc/can_gain_stacks()
+	return owner.stat != DEAD
+
+/datum/status_effect/stacking/tick()
+	if(!can_have_status())
+		qdel(src)
+	else
```
Comment (Rohesie) [inline code/datums/status_effects/status_effect.dm]:
```suggestion
	else
```
return

File: code/datums/status_effects/status_effect.dm [hunk 19]
```diff
@@ -0,0 +1,243 @@
...
+		if(stacks_added > 0)
+			tick_interval += delay_before_decay //refreshes time until decay
+		stacks = min(stacks, max_stacks)
+		status_overlay.icon_state = "[overlay_state][stacks]"
+		status_underlay.icon_state = "[underlay_state][stacks]"
+		owner.add_overlay(status_overlay)
+		owner.underlays += status_underlay
+	else
+		fadeout_effect() 
+		qdel(src) //deletes status if stacks fall under one
```
Comment (Rohesie) [inline code/datums/status_effects/status_effect.dm]: Early return.

File: code/datums/status_effects/status_effect.dm [hunk 20]
```diff
@@ -0,0 +1,243 @@
...
+
+/datum/status_effect/stacking/on_creation(mob/living/new_owner, stacks_to_apply)
+	..()
+	src.add_stacks(stacks_to_apply)
```
Comment (Rohesie) [inline code/datums/status_effects/status_effect.dm]:
```suggestion
	add_stacks(stacks_to_apply)
```

File: code/modules/mob/living/status_procs.dm [hunk 21]
```diff
@@ -1,70 +1,312 @@
...
-	return
-
-/mob/living/proc/set_knocked_out(amount)
-	if(status_flags & CANKNOCKOUT)
-		knocked_out = max(amount,0)
-		update_canmove()
-	return
+	if(((status_flags & CANSTUN) && !HAS_TRAIT(src, TRAIT_STUNIMMUNE)) || ignore_canstun)
+		if(absorb_stun(amount, ignore_canstun))
+			return
```
Comment (Rohesie) [inline code/modules/mob/living/status_procs.dm]: Early return.

File: code/modules/mob/living/status_procs.dm [hunk 22]
```diff
@@ -1,70 +1,312 @@
...
+		else if(amount > 0)
+			S = apply_status_effect(STATUS_EFFECT_STUN, amount, updating)
+		return S
+
+/mob/living/proc/SetStun(amount, updating = TRUE, ignore_canstun = FALSE) //Sets remaining duration
+	if(SEND_SIGNAL(src, COMSIG_LIVING_STATUS_STUN, amount, updating, ignore_canstun) & COMPONENT_NO_STUN)
+		return
+	if(((status_flags & CANSTUN) && !HAS_TRAIT(src, TRAIT_STUNIMMUNE)) || ignore_canstun)
+		var/datum/status_effect/incapacitating/stun/S = IsStun()
+		if(amount <= 0)
```
Comment (Rohesie) [inline code/modules/mob/living/status_procs.dm]: early return.

File: code/modules/mob/living/status_procs.dm [hunk 23]
```diff
@@ -1,70 +1,312 @@
...
+			else
+				S = apply_status_effect(STATUS_EFFECT_STUN, amount, updating)
+		return S
+
+/mob/living/proc/AdjustStun(amount, updating = TRUE, ignore_canstun = FALSE) //Adds to remaining duration
+	if(SEND_SIGNAL(src, COMSIG_LIVING_STATUS_STUN, amount, updating, ignore_canstun) & COMPONENT_NO_STUN)
+		return
+	if(((status_flags & CANSTUN) && !HAS_TRAIT(src, TRAIT_STUNIMMUNE)) || ignore_canstun)
+		if(absorb_stun(amount, ignore_canstun))
+			return
```
Comment (Rohesie) [inline code/modules/mob/living/status_procs.dm]: Early return.

File: code/modules/mob/living/status_procs.dm [hunk 24]
```diff
@@ -1,70 +1,312 @@
...
+	if(IsKnockdown())
+		return 0
+	return Knockdown(amount, updating, ignore_canstun)
+
+/mob/living/proc/Knockdown(amount, updating = TRUE, ignore_canstun = FALSE) //Can't go below remaining duration
+	if(SEND_SIGNAL(src, COMSIG_LIVING_STATUS_KNOCKDOWN, amount, updating, ignore_canstun) & COMPONENT_NO_STUN)
+		return
+	if(((status_flags & CANKNOCKDOWN) && !HAS_TRAIT(src, TRAIT_STUNIMMUNE)) || ignore_canstun)
+		if(absorb_stun(amount, ignore_canstun))
+			return
```
Comment (Rohesie) [inline code/modules/mob/living/status_procs.dm]: Early return.

File: code/modules/mob/living/status_procs.dm [hunk 25]
```diff
@@ -1,70 +1,312 @@
...
+		else if(amount > 0)
+			K = apply_status_effect(STATUS_EFFECT_KNOCKDOWN, amount, updating)
+		return K
+
+/mob/living/proc/SetKnockdown(amount, updating = TRUE, ignore_canstun = FALSE) //Sets remaining duration
+	if(SEND_SIGNAL(src, COMSIG_LIVING_STATUS_KNOCKDOWN, amount, updating, ignore_canstun) & COMPONENT_NO_STUN)
+		return
+	if(((status_flags & CANKNOCKDOWN) && !HAS_TRAIT(src, TRAIT_STUNIMMUNE)) || ignore_canstun)
+		var/datum/status_effect/incapacitating/knockdown/K = IsKnockdown()
+		if(amount <= 0)
```
Comment (Rohesie) [inline code/modules/mob/living/status_procs.dm]: Early return.

File: code/modules/mob/living/status_procs.dm [hunk 26]
```diff
@@ -1,70 +1,312 @@
...
+			else
+				K = apply_status_effect(STATUS_EFFECT_KNOCKDOWN, amount, updating)
+		return K
+
+/mob/living/proc/AdjustKnockdown(amount, updating = TRUE, ignore_canstun = FALSE) //Adds to remaining duration
+	if(SEND_SIGNAL(src, COMSIG_LIVING_STATUS_KNOCKDOWN, amount, updating, ignore_canstun) & COMPONENT_NO_STUN)
+		return
+	if(((status_flags & CANKNOCKDOWN) && !HAS_TRAIT(src, TRAIT_STUNIMMUNE)) || ignore_canstun)
+		if(absorb_stun(amount, ignore_canstun))
+			return
```
Comment (Rohesie) [inline code/modules/mob/living/status_procs.dm]: Early return.

File: code/modules/mob/living/status_procs.dm [hunk 27]
```diff
@@ -1,70 +1,312 @@
...
+	if(S)
+		return S.duration - world.time
+	return 0
+
+/mob/living/proc/Sleeping(amount, updating = TRUE, ignore_canstun = FALSE) //Can't go below remaining duration
+	if(SEND_SIGNAL(src, COMSIG_LIVING_STATUS_SLEEP, amount, updating, ignore_canstun) & COMPONENT_NO_STUN)
+		return
+	if((!HAS_TRAIT(src, TRAIT_SLEEPIMMUNE)) || ignore_canstun)
+		var/datum/status_effect/incapacitating/sleeping/S = IsSleeping()
+		if(S)
```
Comment (Rohesie) [inline code/modules/mob/living/status_procs.dm]: Early return.

File: code/modules/mob/living/status_procs.dm [hunk 28]
```diff
@@ -1,70 +1,312 @@
...
+		else if(amount > 0)
+			S = apply_status_effect(STATUS_EFFECT_SLEEPING, amount, updating)
+		return S
+
+/mob/living/proc/SetSleeping(amount, updating = TRUE, ignore_canstun = FALSE) //Sets remaining duration
+	if(SEND_SIGNAL(src, COMSIG_LIVING_STATUS_SLEEP, amount, updating, ignore_canstun) & COMPONENT_NO_STUN)
+		return
+	if((!HAS_TRAIT(src, TRAIT_SLEEPIMMUNE)) || ignore_canstun)
+		var/datum/status_effect/incapacitating/sleeping/S = IsSleeping()
+		if(amount <= 0)
```
Comment (Rohesie) [inline code/modules/mob/living/status_procs.dm]: Early return.

File: code/modules/mob/living/status_procs.dm [hunk 29]
```diff
@@ -1,70 +1,312 @@
...
+		else
+			S = apply_status_effect(STATUS_EFFECT_SLEEPING, amount, updating)
+		return S
+
+/mob/living/proc/AdjustSleeping(amount, updating = TRUE, ignore_canstun = FALSE) //Adds to remaining duration
+	if(SEND_SIGNAL(src, COMSIG_LIVING_STATUS_SLEEP, amount, updating, ignore_canstun) & COMPONENT_NO_STUN)
+		return
+	if((!HAS_TRAIT(src, TRAIT_SLEEPIMMUNE)) || ignore_canstun)
+		var/datum/status_effect/incapacitating/sleeping/S = IsSleeping()
+		if(S)
```
Comment (Rohesie) [inline code/modules/mob/living/status_procs.dm]: Early return.

File: code/modules/mob/living/status_procs.dm [hunk 30]
```diff
@@ -1,70 +1,312 @@
...
+	if(priority_absorb_key)
+		if(amount) //don't spam up the chat for continuous stuns
+			if(priority_absorb_key["visible_message"] || priority_absorb_key["self_message"])
+				if(priority_absorb_key["visible_message"] && priority_absorb_key["self_message"])
+					visible_message("<span class='warning'>[src][priority_absorb_key["visible_message"]]</span>", "<span class='boldwarning'>[priority_absorb_key["self_message"]]</span>")
+				else if(priority_absorb_key["visible_message"])
+					visible_message("<span class='warning'>[src][priority_absorb_key["visible_message"]]</span>")
+				else if(priority_absorb_key["self_message"])
+					to_chat(src, "<span class='boldwarning'>[priority_absorb_key["self_message"]]</span>")
+			priority_absorb_key["stuns_absorbed"] += amount
```
Comment (Rohesie) [inline code/modules/mob/living/status_procs.dm]: This logic could really use some improvement.

**Comment by Rohesie** [comment]
Since every stun is more or less the exact time as described in the code, instead of the time minus a random number between 2 and 0 seconds, stuns have become much more brutal. This will need a bit of re-balancing.

**Pull request merged**